errors.go

```diff
@@ -0,0 +1,120 @@
+package deis
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+)
+
+const (
+	// formatErrUnknown is used to create an dynamic error if no error matches
+	formatErrUnknown = "Unknown Error (%d): %s"
+	// fieldReqMsg is API error stating a field is required.
+	fieldReqMsg    = "This field is required."
+	invalidUserMsg = "Enter a valid username. This value may contain only letters, numbers and @/./+/-/_ characters."
+	failedLoginMsg = "Unable to log in with provided credentials."
+)
+
+var (
+	// ErrNotFound is returned when the server returns a 404
+	ErrNotFound = errors.New("Not Found")
+	// ErrServerError is returned when the server returns a 500
+	ErrServerError = errors.New("Internal Server Error")
+	// ErrMethodNotAllowed is thrown when using a unsupposrted method.
+	// This should not come up unless there in an bug in the SDK.
+	ErrMethodNotAllowed = errors.New("Method Not Allowed")
+	// ErrMissingUsername is returned when a username is not sent with the request.
+	ErrMissingUsername = errors.New("Username is required")
+	// ErrInvalidUsername is returned when the user specifies an invalid username.
+	ErrInvalidUsername = errors.New(invalidUserMsg)
+	// ErrMissingPassword is returned when a password is not sent with the request.
+	ErrMissingPassword = errors.New("Password is required")
+	// ErrParsing is returned when the SDK fails at parsing an error.
+	ErrParsing = errors.New("Unable to Parse Error")
+	// ErrLogin is returned when the api cannot login fails with provided username and password
+	ErrLogin = errors.New(failedLoginMsg)
+)
+
+func checkForErrors(res *http.Response) error {
+	if res.StatusCode >= 200 && res.StatusCode < 400 {
+		return nil
+	}
+
+	// Fix json.Decoder bug in <go1.7
+	defer func() {
+		io.Copy(ioutil.Discard, res.Body)
+		res.Body.Close()
+	}()
+
+	switch res.StatusCode {
+	case 400:
+		bodyMap := make(map[string]interface{})
+		if err := json.NewDecoder(res.Body).Decode(&bodyMap); err != nil {
+			return ErrParsing
```

Comment:
perhaps we should also return the response body so the user can determine the error on their own?